File: src/main/java/com/example/authapp/entity/Product.java
package com.example.authapp.entity;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.OffsetDateTime;
import java.util.HashSet;
import java.util.Set;
import java.util.UUID;

@Entity
@Table(name = "products")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Product {
    @Id
    @GeneratedValue
    @Column(columnDefinition = "UUID", updatable = false, nullable = false)
    private UUID id;

    @Column(nullable = false, unique = true, columnDefinition = "TEXT")
    private String slug;

    @Column(name = "product_name", nullable = false)
    private String productName;

    @Column(name = "brand_name")
    private String brandName;

    @Column(name = "image_url", columnDefinition = "TEXT")
    private String imageUrl;

    private String sku;

    @Column(name = "sale_price", nullable = false)
    private double salePrice;

    @Column(name = "compare_price")
    private Double comparePrice;

    @Column(name = "buying_price")
    private Double buyingPrice;

    @Column(nullable = false)
    private int quantity;

    @Column(name = "short_description", nullable = false, length = 165)
    private String shortDescription;

    @Column(name = "product_description", nullable = false, columnDefinition = "TEXT")
    private String productDescription;

    @Column(name = "product_type", length = 64)
    private String productType;

    @Builder.Default
    private boolean published = false;

    @Column(name = "disable_out_of_stock")
    @Builder.Default
    private boolean disableOutOfStock = true;

    @Column(name = "rating_average")
    private Double ratingAverage;

    @Column(name = "review_count")
    private Integer reviewCount;

    @Column(columnDefinition = "TEXT")
    private String note;

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private OffsetDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at", nullable = false)
    private OffsetDateTime updatedAt;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "created_by")
    private StaffAccount createdBy;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "updated_by")
    private StaffAccount updatedBy;

    @com.fasterxml.jackson.annotation.JsonProperty(value = "tags", access = com.fasterxml.jackson.annotation.JsonProperty.Access.READ_ONLY)
    @ManyToMany(fetch = FetchType.LAZY)
    @JoinTable(
        name = "product_tags",
        joinColumns = @JoinColumn(name = "product_id"),
        inverseJoinColumns = @JoinColumn(name = "tag_id")
    )
    @Builder.Default
    private Set<Tag> tags = new HashSet<>();

    @Transient
    @com.fasterxml.jackson.annotation.JsonProperty(value = "tags", access = com.fasterxml.jackson.annotation.JsonProperty.Access.WRITE_ONLY)
    private Set<String> tagNames;

    @ElementCollection(fetch = FetchType.EAGER)
    @CollectionTable(name = "product_sizes", joinColumns = @JoinColumn(name = "product_id"))
    @Column(name = "size")
    @Builder.Default
    private Set<String> sizes = new HashSet<>();

    @ElementCollection(fetch = FetchType.EAGER)
    @CollectionTable(name = "product_colors", joinColumns = @JoinColumn(name = "product_id"))
    @Column(name = "color")
    @Builder.Default
    private Set<String> colors = new HashSet<>();

    @ManyToMany(fetch = FetchType.LAZY)
    @JoinTable(
        name = "product_categories",
        joinColumns = @JoinColumn(name = "product_id"),
        inverseJoinColumns = @JoinColumn(name = "category_id")
    )
    @Builder.Default
    private Set<Category> categories = new HashSet<>();

    @OneToMany(mappedBy = "product", fetch = FetchType.LAZY, cascade = CascadeType.ALL)
    @Builder.Default
    private Set<Gallery> gallery = new HashSet<>();

    @Transient
    private Set<UUID> categoryIds;

    public Set<UUID> getCategoryIds() {
        if (categories == null) return new HashSet<>();
        return categories.stream().map(Category::getId).collect(java.util.stream.Collectors.toSet());
    }

    public void setCategoryIds(Set<UUID> categoryIds) {
        this.categoryIds = categoryIds;
    }
}

File: LeNhatTruong/lib/models/product.dart
import 'tag.dart';

class Product {
  final String id;
  final String slug;
  final String productName;
  final String? brandName;
  final String? imageUrl;
  final String? sku;
  final double salePrice;
  final double? comparePrice;
  final double? buyingPrice;
  final int quantity;
  final String shortDescription;
  final String productDescription;
  final String? productType;
  final bool published;
  final bool disableOutOfStock;
  final double? ratingAverage;
  final int? reviewCount;
  final String? note;
  final List<Tag> tags;

  Product({
    required this.id,
    required this.slug,
    required this.productName,
    this.brandName,
    this.imageUrl,
    this.sku,
    required this.salePrice,
    this.comparePrice,
    this.buyingPrice,
    required this.quantity,
    required this.shortDescription,
    required this.productDescription,
    this.productType,
    required this.published,
    required this.disableOutOfStock,
    this.ratingAverage,
    this.reviewCount,
    this.note,
    required this.tags,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['id'] ?? '',
      slug: json['slug'] ?? '',
      productName: json['productName'] ?? '',
      brandName: json['brandName'],
      imageUrl: json['imageUrl'],
      sku: json['sku'],
      salePrice: (json['salePrice'] as num?)?.toDouble() ?? 0.0,
      comparePrice: (json['comparePrice'] as num?)?.toDouble(),
      buyingPrice: (json['buyingPrice'] as num?)?.toDouble(),
      quantity: json['quantity'] ?? 0,
      shortDescription: json['shortDescription'] ?? '',
      productDescription: json['productDescription'] ?? '',
      productType: json['productType'],
      published: json['published'] ?? false,
      disableOutOfStock: json['disableOutOfStock'] ?? true,
      ratingAverage: (json['ratingAverage'] as num?)?.toDouble(),
      reviewCount: json['reviewCount'],
      note: json['note'],
      tags: (json['tags'] as List?)
              ?.map((item) => Tag.fromJson(item))
              .toList() ??
          [],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'slug': slug,
      'productName': productName,
      'brandName': brandName,
      'imageUrl': imageUrl,
      'sku': sku,
      'salePrice': salePrice,
      'comparePrice': comparePrice,
      'buyingPrice': buyingPrice,
      'quantity': quantity,
      'shortDescription': shortDescription,
      'productDescription': productDescription,
      'productType': productType,
      'published': published,
      'disableOutOfStock': disableOutOfStock,
      'ratingAverage': ratingAverage,
      'reviewCount': reviewCount,
      'note': note,
      'tags': tags.map((t) => t.toJson()).toList(),
    };
  }
}
